src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedList`1.cs

```diff
@@ -0,0 +1,794 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using IEnumerable = System.Collections.IEnumerable;
+using IEnumerator = System.Collections.IEnumerator;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    internal interface IByRefComparer<T>
+    {
+        int Compare(in T first, in T second);
+    }
+
+    /// <summary>
+    /// Segmented list implementation, copied from Microsoft.Exchange.Collections.
+    /// </summary>
+    /// <typeparam name="T">The type of the list element.</typeparam>
+    /// <remarks>
+    /// This class implement a list which is allocated in segments, to avoid large lists to go into LOH.
+    /// </remarks>
+    internal sealed class SegmentedList<T> : ICollection<T>, IReadOnlyList<T>
```

Comment:
super surprising that this a mutable list, but doesn't implement IList<T>

Reply:
i would also be ok with this not implementing *any* interfaces.  it would be nice to ensure that things like iteration are alloc-free and that sort of thing.